pallets/liquidity-mining/src/lib.rs

```diff
@@ -0,0 +1,1861 @@
+// This file is part of HydraDX
+
+// Copyright (C) 2020-2021  Intergalactic, Limited (GIB).
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Abbr:
+//  rpvs - reward per valued share
+//  rpz - reward per share in global pool
+
+// Notion spec naming map:
+// * shares                 -> s
+// * total_shares           -> S
+// * valued_shares          -> s'
+// * total_valued_shares    -> S'
+// * stake_in_global_pool   -> z
+// * total_shares_z         -> Z
+// * multiplier             -> m
+
+//! # Liquidity mining pallet
+//!
+//! ## Overview
+//!
+//! This pallet provide functionality for liquidity mining program with time incentive(loyalty
+//! factor).
+//! Users are rewarded for each period they stay in liq. mining program.
+//!
+//! Reward per one period is derived from the user's loyalty factor which grows with time(periods)
+//! the user is in the liq. mining and amount of LP shares user locked into deposit.
+//! User's loyalty factor is reseted if the user exits and reenters liq. mining pool.
+//! User can claim rewards without penalties(loyalty factor is no reseted), only withdrawing shares
+//! is penalized by loyalty factor reset.
+//! User is rewarded from the next period he enters.
+//!
+//! User's deposit in liq. mining pool is represented by NFT which is minted for the user when he
+//! enters liq. mining and is burned when he exits. NFT representing deposit is tradable.
+
+// TODO:
+// add to docs blocks are related to relay chain blocks
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![allow(clippy::unused_unit)]
+#![allow(clippy::upper_case_acronyms)]
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+pub mod migration;
+pub mod weights;
+
+pub use pallet::*;
+
+type PoolId = u32;
+type GlobalPoolId = PoolId;
+type PoolMultiplier = FixedU128;
+
+//This value is result of: u128::from_le_bytes([255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 0, 0, 0, 0])
```

Comment:
Shouldn't this be handled by NFT pallet itself? (constants in the NFT pallet?)

Reply:
Ok now I see that not but could use better writeup of how this came to be and why.